www/_blog/2022-08-11-authentication-tutorial.mdx

```diff
@@ -0,0 +1,516 @@
+
+We can pull the cookie value using this spaghetti of iterators and options
+
+```rust
+let session_token = req
+    .headers()
+    .get_all("Cookie")
+    .iter()
+    .filter_map(|cookie| {
+        cookie
+            .to_str()
+            .ok()
+            .and_then(|cookie| cookie.parse::<cookie::Cookie>().ok())
+    })
+    .find_map(|cookie| {
+        (cookie.name() == USER_COOKIE_NAME).then(move || cookie.value().to_owned())
+    })
+    .and_then(|cookie_value| cookie_value.parse::<u128>().ok());
+```
+
+With our value we now need to make a query.
+
+We can do that using a similar way to how we would do in code.
+
+```sql
+SELECT id, username 
+FROM users 
+WHERE id = (SELECT user_id FROM sessions WHERE session_token = $1);
+```
+
+Or we can use SQL's joins
+
+```sql
+SELECT id, username 
+FROM users 
+JOIN sessions ON user_id = id 
+WHERE session_token = $1;
+```
+
+##### Auth middleware
+
+[Callback to last post we spoke a bit about middleware. You can read more about middleware there](https://www.shuttle.rs/blog/2022/08/04/middleware). 
+
+In our middleware we will get a little fancy and make the user pulling lazy. This is so that requests that don't need user data don't have to make a database trip. Rather than adding our user straight onto the the request we split things apart. We first create a `AuthState` which contains the session token, the database and a placeholder for our user `(Option<User>)`
+
+```rust
+#[derive(Clone)]
+pub(crate) struct AuthState(Option<(u128, Option<User>, Database)>);
+
+pub(crate) async fn auth<B>(
+    mut req: http::Request<B>,
+    next: axum::middleware::Next<B>,
+    database: Database,
+) -> axum::response::Response {
+    let session_token = /* cookie logic */;
+
+    req.extensions_mut()
+        .insert(AuthState(session_token.map(|v| (v, None, database))));
+
+    next.run(req).await
+}
+```
+
+Then we create a method on `AuthState` which actually makes the database request.
+
+```rust
+pub async fn get_user(&mut self) -> Option<&User> {
+    let (session_token, store, database) = self.0.as_mut()?;
+    if store.is_none() {
+        const QUERY: &str =
+            "SELECT id, username FROM users JOIN sessions ON user_id = id WHERE session_token = $1;";
+
+        let user: Option<(i32, String)> = sqlx::query_as(QUERY)
+            .bind(&session_token.into_database_value())
+            .fetch_optional(&*database)
+            .await
+            .unwrap();
+
+        if let Some((_id, username)) = user {
+            *store = Some(User { username });
+        }
+    }
+    store.as_ref()
+}
+```
+
+Here because we store the user internally it means that if another middleware gets the user and then a handler gets the user it results in one database request not two!
+
+[The full demo includes registering the middleware](https://github.com/kaleidawave/axum-shuttle-postgres-authentication-demo/blob/ba71a914055f312636581f5e82172b1078e7b9eb/src/lib.rs#L53-L68).

```

Comment:
Maybe we want to include the registration code at the minimum to complete this post?